docs/aot-jit-switch.md

```diff
@@ -0,0 +1,184 @@
+ R4.0-AOT-cost   | cost estimation using v4.0 Ahead-Of-Time costing implementation
+ R4.0-AOT-verify | spending condition verification using v4.0 Ahead-Of-Time interpreter implementation
+ R5.0-JIT-verify | spending condition verification using v5.0 simplified interpreter with Just-In-Time costing of fullReduction and AOT sigma protocol costing.
+ skip-pool-tx    | skip pool transaction when building a new block candidate 
+ skip-accept     | skip script evaluation (both costing and verification) and treat it as True proposition (accept spending) 
+ skip-reject     | skip script evaluation (both costing and verification) and treat it as False proposition (reject spending) 
+ Validation Context | a tuple of (`Block Type`, `SF Status`, `Script Version`)
+ SF Status       | soft-fork status of the block. The status is `active` when enough votes have been collected.
+
+### Script Validation Rules Summary
+
+Validation of scripts in blocks is defined for each release and depend on _validation
+context_ which includes type of block, soft-fork status and script version. We denote
+blocks being created by miners as `candidate` and those distributed across network as
+`mined`.
+
+Thus, we have 8 different validation contexts multiplied by 2 node versions
+having in total 16 validation rules as summarized in the following table, which 
+specifies the action a node have to take in the given contexts.
+
+Rule#| SF Status| Block Type| Script Version | Release | Validation Action 
+-----|----------|-----------|----------------|---------|--------
+1    | inactive | candidate | Script v1 | v4.0 | R4.0-AOT-cost, R4.0-AOT-verify
+2    | inactive | candidate | Script v1 | v5.0 | R4.0-AOT-cost, R4.0-AOT-verify
+3    | inactive | candidate | Script v2 | v4.0 | skip-pool-tx (cannot handle)
+4    | inactive | candidate | Script v2 | v5.0 | skip-pool-tx (wait activation)
+||||
+5    | inactive | mined     | Script v1 | v4.0 | R4.0-AOT-cost, R4.0-AOT-verify 
+6    | inactive | mined     | Script v1 | v5.0 | R4.0-AOT-cost, R4.0-AOT-verify 
+7    | inactive | mined     | Script v2 | v4.0 | skip-reject (cannot handle)
+8    | inactive | mined     | Script v2 | v5.0 | skip-reject (wait activation)
+||||
+9    | active   | candidate | Script v1 | v4.0 | R4.0-AOT-cost, R4.0-AOT-verify
+10   | active   | candidate | Script v1 | v5.0 | R5.0-JIT-verify 
+11   | active   | candidate | Script v2 | v4.0 | skip-pool-tx (cannot handle)
+12   | active   | candidate | Script v2 | v5.0 | R5.0-JIT-verify 
+||||
+13   | active   | mined     | Script v1 | v4.0 | skip-accept (rely on majority)
+14   | active   | mined     | Script v1 | v5.0 | R5.0-JIT-verify 
+15   | active   | mined     | Script v2 | v4.0 | skip-accept (rely on majority)
+16   | active   | mined     | Script v2 | v5.0 | R5.0-JIT-verify 
+
+Note the following properties of the validation rules.
+
+1. Rules 1-4 specify creation of new candidate blocks _before_ soft-fork is activated. 
```

Comment:
Please note that block creation is not a part of the protocol, and miners can do whatever they want here. Even more, possibly some miners are doing custom block assembly right now. 

This it would be reasonable to split consensus and non-consensus procedures, by e.g. moving block assembly rules into an appendix.

Reply:
Yes, but on the other hand it may be easier to look at the whole set of rules in one table. I will move you comment in the test and make this point more clear.